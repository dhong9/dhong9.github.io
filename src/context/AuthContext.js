/**
=========================================================
* Danyo-1.2
=========================================================

* Copyright 2024 Danyo (https://www.danyo.tech)

Coded by www.danyo.tech

=========================================================

* The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
*/

import { createContext, useState, useEffect } from "react";
import { googleLogout, useGoogleLogin } from "@react-oauth/google";
import { useNavigate } from "react-router-dom";
import axios from "axios";

// Services
import {
  addAccount,
  getUser,
  loginAccount,
  updateAccount,
  deleteAccount,
} from "services/accountsService";
import { getGoogleUser } from "services/googleService";

// prop-types is a library for typechecking of props
import PropTypes from "prop-types";

const AuthContext = createContext();

export default AuthContext;

export const AuthProvider = ({ children }) => {
  // Getting auth token priority
  // 1. Session storage
  // 2. Local storage
  const sessionToken = sessionStorage.getItem("authTokens");
  const localToken = localStorage.getItem("authTokens");
  const [authTokens, setAuthTokens] = useState(() => sessionToken || localToken || null);
  const [user, setUser] = useState(null);
  const [profile, setProfile] = useState(null);

  const history = useNavigate();

  /**
   * Logs in a user
   * @param {string} username username or email address
   * @param {string} password password
   * @param {boolean} rememberMe flag to remember user or not
   * @param {Function} success success callback
   * @param {Function} error error callback
   */
  const loginUser = (username, password, rememberMe, success, error) => {
    loginAccount(
      /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/.test(username)
        ? { email: username, password }
        : { username, password },
      (response) => {
        // Unpack token data
        const { auth_token } = response.data;
        axios.defaults.headers.common["Authorization"] = "Token " + auth_token;

        // If user wants to be remembered,
        // save their token in a local cache
        if (rememberMe) {
          localStorage.setItem("authTokens", auth_token);
        }
        // If not,
        // just maintain their token until they close their session
        else {
          sessionStorage.setItem("authTokens", auth_token);
        }

        // Navigate to home page and show success status
        history("/");
        success();

        // Get user data
        getUser(({ data }) => {
          // data {
          //     email: ...,
          //     username: ...
          // }
          if (rememberMe) {
            localStorage.setItem("user", data);
          } else {
            sessionStorage.setItem("user", data);
          }
        }, console.error);
      },
      error
    );
  };

  /**
   * Adds user to database
   * @param {string} email email address
   * @param {string} username username
   * @param {string} password password
   * @param {Function} success success callback
   * @param {Function} error error callback
   */
  const registerUser = (email, username, password, success, error) => {
    addAccount(
      {
        email,
        username,
        password,
      },
      () => {
        // If the user successfully registered,
        // take them so sign in page and display success message
        history("/pages/authentication/sign-in");
        success();
      },
      error
    );
  };

  const updateUser = (id, email, username, password, password2, success, error) => {
    updateAccount(
      id,
      {
        email,
        username,
        password,
        password2,
      },
      success,
      error
    );
  };

  const logoutUser = () => {
    googleLogout();

    // Cleared saved data
    localStorage.removeItem("authTokens");
    sessionStorage.removeItem("authTokens");
    localStorage.removeItem("user");
    sessionStorage.removeItem("user");
    setAuthTokens(null);
    setUser(null);
    setProfile(null);

    // Go back to home page
    history("/");
  };

  const deleteUser = (id, success, error) => {
    deleteAccount(id, success, error);
  };

  // Google login functions
  const googleLogin = useGoogleLogin({
    onSuccess: (codeResponse) => {
      // data has access and refresh tokens
      setAuthTokens(codeResponse);
      setUser(codeResponse);
      sessionStorage.setItem("authTokens", JSON.stringify(codeResponse));
      history("/");
    },
    onError: (error) => {
      console.error(error);
    },
  });

  const contextData = {
    user,
    setUser,
    authTokens,
    setAuthTokens,
    registerUser,
    loginUser,
    logoutUser,
    updateUser,
    deleteUser,
    profile,
    googleLogin,
  };

  useEffect(() => {
    // Get Google user
    if (authTokens && authTokens.access_token) {
      getGoogleUser(
        authTokens.access_token,
        (res) => {
          setProfile(res.data);
        },
        console.error
      );
    }
  }, [authTokens]);

  return <AuthContext.Provider value={contextData}>{children}</AuthContext.Provider>;
};

// Typechecking props of AuthContext
AuthProvider.propTypes = {
  children: PropTypes.node,
};

AuthProvider.defaultProps = {
  children: null,
};
