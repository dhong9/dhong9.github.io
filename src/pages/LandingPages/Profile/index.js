import { useContext } from "react";

// @mui material icons
import AccountCircleIcon from "@mui/icons-material/AccountCircle";

// Material Kit 2 React examples
import DefaultNavbar from "examples/Navbars/DefaultNavbar";

// Routes
import routes from "routes";

// Authentication
import SignOut from "layouts/pages/authentication/sign-out";
import AuthContext from "context/AuthContext";

function Profile() {
  let { user } = useContext(AuthContext);

  const accountObj = {
    name: user.username,
    icon: <AccountCircleIcon />,
    collapse: [
      {
        name: "sign out",
        route: "/pages/authentication/sign-out",
        component: <SignOut />,
      },
    ],
  };

  return (
    <>
      <DefaultNavbar routes={[...routes, accountObj]} sticky />
      <h3>My Profile</h3>
    </>
  );
}

export default Profile;
