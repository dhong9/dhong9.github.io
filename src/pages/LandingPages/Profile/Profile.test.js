// React testing libraries
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

// Material Kit 2 React themes
import { ThemeProvider } from "@mui/material/styles";
import theme from "assets/theme";

// Authentication
import axios from "axios";
import AuthContext, { AuthProvider } from "context/AuthContext";
import jwtDecode from "jwt-decode";

// Axios
import MockAdapter from "axios-mock-adapter";

// Component to test
import Profile from "pages/LandingPages/Profile";

jest.mock("services/baseService", () => ({
  getRequest: jest.fn(),
  putRequest: jest.fn(),
  postRequest: jest.fn(),
}));

// Setup axios mock
const mock = new MockAdapter(axios);

mock.onPost("accounts/update/").reply(200, {});

// Define Mocks
jest.mock("jwt-decode");
jest.mock("react-monaco-editor", () => {
  const { forwardRef } = jest.requireActual("react");
  return {
    __esModule: true,
    default: forwardRef(() => <div>Mock Editor</div>),
  };
});
jest.mock("draft-convert", () => {
  return {
    convertFromHTML: jest.fn(),
    convertToRaw: jest.fn(),
  };
});
jest.mock("react-router-dom", () => ({
  Link: jest.fn(({ to, children }) => <a href={to}>{children}</a>),
  useNavigate: jest.fn,
}));
jest.mock("examples/Navbars/DefaultNavbar", () => {
  const { forwardRef } = jest.requireActual("react");
  return {
    __esModule: true,
    default: forwardRef(() => <div>Mock Navbar</div>),
  };
});

describe("Profile", () => {
  beforeEach(() => {
    localStorage.clear();
  });

  afterEach(() => {
    localStorage.removeItem("authTokens");
    jest.clearAllMocks();
  });

  it("updates user profile", () => {
    // Mock tokens
    const mockToken = "mocked_jwt_value";
    const refreshToken = "mocked_refresh_value";

    localStorage.setItem(
      "authTokens",
      JSON.stringify({ access: mockToken, refresh: refreshToken })
    );

    // Set a mock payload for the decoded token
    const mockPayload = { user: "John Doe", exp: 1893456000 };
    jwtDecode.mockReturnValue(mockPayload);

    const contextData = {
      loginUser: jest.fn(),
      user: {
        username: "giri"
      }
    };
    const { getByLabelText, getByText, queryByText } = render(
      <AuthContext.Provider value={contextData}>
        <AuthProvider>
          <ThemeProvider theme={theme}>
            <Profile />
          </ThemeProvider>
        </AuthProvider>
      </AuthContext.Provider>
    );

    // Get form elements
    const usernameInput = getByLabelText("Username");
    const emailInput = getByLabelText("Email");
    const passwordInput = getByLabelText("Password");
    const password2Input = getByLabelText("Confirm Password");
    const saveButton = getByText("save changes");

    fireEvent.change(usernameInput, { target: { value: "" } });
    fireEvent.change(emailInput, { target: { value: "" } });

    // Sign up with no form input
    fireEvent.click(saveButton);

    // Put data into the form
    fireEvent.change(usernameInput, { target: { value: "expertTester" } });
    fireEvent.change(emailInput, { target: { value: "expertTester@aol.com" } });
    fireEvent.change(passwordInput, { target: { value: "validPassword" } });
    fireEvent.change(password2Input, { target: { value: "validPassword" } });

    // Resubmit data
    fireEvent.click(saveButton);

    // Errors should clear
    expect(queryByText("Username is required.")).not.toBeInTheDocument();
    expect(queryByText("Password is required.")).not.toBeInTheDocument();
    expect(queryByText("Password confirmation is required.")).not.toBeInTheDocument();
  });
});
