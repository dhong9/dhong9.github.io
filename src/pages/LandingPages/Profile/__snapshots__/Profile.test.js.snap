// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders 1`] = `
[
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-nqeyix-MuiContainer-root"
  >
    <div
      className="MuiBox-root css-1yv1dw1"
    >
      <div
        className="MuiBox-root css-1dyncq9"
      >
        <a
          href="/"
        >
          <span
            className="MuiTypography-root MuiTypography-button css-lan4mq-MuiTypography-root"
          >
            Daniel Hong Portfolio
          </span>
        </a>
        <div
          className="MuiBox-root css-k0f0w7"
        >
          <a
            href="/doughnutRider"
          >
            <p
              className="MuiTypography-root MuiTypography-body2 css-incl6n-MuiTypography-root"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-eimhud-MuiSvgIcon-root"
                data-testid="DonutSmallIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 9.16V2c-5 .5-9 4.79-9 10s4 9.5 9 10v-7.16c-1-.41-2-1.52-2-2.84s1-2.43 2-2.84zM14.86 11H22c-.48-4.75-4-8.53-9-9v7.16c1 .3 1.52.98 1.86 1.84zM13 14.84V22c5-.47 8.52-4.25 9-9h-7.14c-.34.86-.86 1.54-1.86 1.84z"
                />
              </svg>
            </p>
            <span
              className="MuiTypography-root MuiTypography-button css-129xe1v-MuiTypography-root"
            >
              doughnut rider
            </span>
            <p
              className="MuiTypography-root MuiTypography-body2 css-wnhb9o-MuiTypography-root"
            >
              <span
                aria-hidden={true}
                className="material-icons-round notranslate MuiIcon-root MuiIcon-fontSizeInherit css-x5zaln-MuiIcon-root"
              />
            </p>
          </a>
          <div
            className="MuiBox-root css-111f5ru"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <p
              className="MuiTypography-root MuiTypography-body2 css-incl6n-MuiTypography-root"
            >
              <span
                aria-hidden={true}
                className="material-icons-round notranslate MuiIcon-root MuiIcon-fontSizeInherit css-1neevca-MuiIcon-root"
              >
                view_day
              </span>
            </p>
            <span
              className="MuiTypography-root MuiTypography-button css-129xe1v-MuiTypography-root"
            >
              sections
            </span>
            <p
              className="MuiTypography-root MuiTypography-body2 css-wnhb9o-MuiTypography-root"
            >
              <span
                aria-hidden={true}
                className="material-icons-round notranslate MuiIcon-root MuiIcon-fontSizeInherit css-x5zaln-MuiIcon-root"
              >
                keyboard_arrow_down
              </span>
            </p>
          </div>
          <a
            className="MuiBox-root css-111f5ru"
            href="https://github.com/dhong9/dhong9.github.io"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            rel="noreferrer"
            target="_blank"
          >
            <p
              className="MuiTypography-root MuiTypography-body2 css-incl6n-MuiTypography-root"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-eimhud-MuiSvgIcon-root"
                data-testid="GitHubIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
                />
              </svg>
            </p>
            <span
              className="MuiTypography-root MuiTypography-button css-129xe1v-MuiTypography-root"
            >
              github
            </span>
            <p
              className="MuiTypography-root MuiTypography-body2 css-wnhb9o-MuiTypography-root"
            >
              <span
                aria-hidden={true}
                className="material-icons-round notranslate MuiIcon-root MuiIcon-fontSizeInherit css-x5zaln-MuiIcon-root"
              />
            </p>
          </a>
          <div
            className="MuiBox-root css-111f5ru"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <p
              className="MuiTypography-root MuiTypography-body2 css-incl6n-MuiTypography-root"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-eimhud-MuiSvgIcon-root"
                data-testid="AccountCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
                />
              </svg>
            </p>
            <span
              className="MuiTypography-root MuiTypography-button css-129xe1v-MuiTypography-root"
            />
            <p
              className="MuiTypography-root MuiTypography-body2 css-wnhb9o-MuiTypography-root"
            >
              <span
                aria-hidden={true}
                className="material-icons-round notranslate MuiIcon-root MuiIcon-fontSizeInherit css-x5zaln-MuiIcon-root"
              >
                keyboard_arrow_down
              </span>
            </p>
          </div>
        </div>
        <div
          className="MuiBox-root css-13l63k6"
        />
        <div
          className="MuiBox-root css-1kcb7bo"
          onClick={[Function]}
        >
          <span
            aria-hidden={true}
            className="material-icons-round notranslate MuiIcon-root MuiIcon-fontSizeDefault css-75zff6-MuiIcon-root"
          >
            menu
          </span>
        </div>
      </div>
      <div
        className="MuiBox-root css-1kfwhtq"
      />
    </div>
  </div>,
  <h3>
    My Profile
  </h3>,
]
`;
